### ISSUE TITLE
proposal: `move_color_to_decoration`

### ISSUE THREAD

**Issue by minhqdao** [dart-lang/linter#3063]
## `move_color_to_decoration`

##Description

Don't provide `Container` with both non-null `color` and `decoration`. Place `color` inside `decoration` instead.

## Details

When a `Container` is provided with both `color` and `decoration` (both non-null), an error is thrown at runtime, and there is currently nothing that warns you about that beforehand. This rule prevents you from running into the error by prompting to move `color` into `decoration`.

## Kind

Guard against errors.

## Good Examples 

```dart
Widget buildArea() {
  return Container(
    decoration: BoxDecoration(
      border: Border.all(color: Colors.white),
      color: Colors.black,
    ),
  );
}
```

## Bad Examples

```dart
Widget buildArea() {
  return Container(
    color: Colors.black,
    decoration: BoxDecoration(
      border: Border.all(color: Colors.white),
    ),
  );
}
```

## Discussion

Those who already ran into the error will know, but those who haven't will certainly do, and that's simply annoying. No one wants to dig through unexpected error messages, even if this one is rather easy to identify.

### Discussion checklist

- [x] List any existing rules this proposal complements, overlaps or conflict with.
Slightly complementing: https://github.com/dart-lang/linter/issues/3060 https://github.com/dart-lang/linter/pull/3061
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
Not found.
- [x] If there's any prior art (e.g., in other linters), please add references here.
None.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn’t any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/master/doc/WritingLints.MD
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo]
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by goderbauer** [comment]
This seems useful to inform people of an illegal configuration at code write time instead of compile time.

**Comment by HansMuller** [comment]
I agree, if we can flag this problem early it's a win for developers.

**Comment by pq** [comment]
This is terrific and addresses a problem that really trips folks up. I looked at this a while ago and recall there are at least a handful of other Flutter classes w/ similar constraints.  Doing a quick scan now I see similar patterns in _at least_: `CupertinoSearchTextField`, `Ink`, `TextStyle`, and `AnimatedContainer`.  Given that, it'd be great to revisit a more general and robust solution.

Using an annotation, for example, we could capture this constraint at the API level.

Assuming we had an `@OnlyOneOf` for instance, we could update `flutter/lib/src/widgets/container.dart` like this:

```dart
  Container({
    Key key,
    this.alignment,
    this.padding,
    @OnlyOneOf(‘decorationGroup’)
    this.color,
    @OnlyOneOf(‘decorationGroup’)
    this.decoration,
    this.foregroundDecoration,
    double width,
    double height,
    BoxConstraints constraints,
    this.margin,
    this.transform,
    this.child,
    this.clipBehavior = Clip.none,
  }) ...
```

with an annotation in `package:meta` like this:

```dart
/// Used to annotate named parameters `p1..pn` in a method or function `f`.
/// Indicates that an invocation of `f` can contain at most one argument
/// corresponding to one of the named parameters `p1..pn`.
///
/// Optionally annotated parameters can be assigned to named groups which
/// allow for more than one mutually exclusive set of parameters in one
/// method or function `f`.  These group ids can also be used for
/// documentation purposes.
class OnlyOneOf {
  /// An id that defines a set of exclusive parameters.
  final String groupId;

  /// Initialize a newly created instance with the given [groupId].
  const OnlyOneOf([this.groupId = '']);
}
```

/fyi @bwilkerson @InMatrix @leafpetersen @munificent who were in on the original conversation about annotating exclusive params vs. a language solution like overloading.

/fyi also @natebosch @jakemac53 @lrhn @eernstg for more language team perspective

**Comment by bwilkerson** [comment]
In terms of using an annotation for this, I think the right next step is to create a new issue (in the sdk repo) to discuss the kinds of information we'd need to capture for the use cases we know about, and then the right set of annotations with which to capture that information. It might well be that `OnlyOneOf` would meet all of the known needs, but I'd like to know that before we make a decision.
